docs/auth/oauth2proxy/provider.md

```diff
@@ -0,0 +1,119 @@
+---
+id: provider
+title: OAuth 2 Proxy Provider
+sidebar_label: OAuth 2 Custom Proxy
+description: Adding OAuth2Proxy as an authentication provider in Backstage
+---
+
+The Backstage `core-plugin-api` package comes with an Oauth2Proxy authentication
+provider that can authenticate users by using a
+[oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy) in front of an
+actual Backstage instance. This enables to reuse existing authentications within
+a cluster. In general the `oauth2-proxy` supports all OpenID Connect providers,
+for more details check this
+[list of supported providers](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider).
+
+## Configuration
+
+The provider configuration can be added to your `app-config.yaml` under the root
+`auth` configuration:
+
+```yaml
+auth:
+  environment: development
+  providers:
+    oauth2proxy: {}
+```
+
+Right now no configuration options are supported. To make use of the provider,
+make sure that the `X-OAUTH2-PROXY-ID-TOKEN` header is set correctly from your
+`oauth2-proxy` instance.
+
+## Adding the provider to the Backstage backend
+
+When using `oauth2proxy` auth you can configure it as described
+[here](https://backstage.io/docs/auth/identity-resolver).
+
+- replace the content of `packages/backend/plugin/auth.ts` with the below and
+  tweak it according to your needs
+
+```ts
+import { Router } from 'express';
+import { PluginEnvironment } from '../types';
+import {
+  createRouter,
+  BackstageIdentity,
+} from '@backstage/plugin-auth-backend';
+import { Router } from 'express';
+import { createOauth2ProxyProvider } from '@backstage/plugin-auth-backend';
+
+interface JWTPayload {
+  email: string;
+}
+
+export const authPlugin = async function createPlugin({
+  logger,
+  database,
+  config,
+  discovery,
+}: PluginEnvironment): Promise<Router> {
+  return await createRouter({
+    logger,
+    config,
+    database,
+    discovery,
+    providerFactories: {
```

Comment:
Perhaps we could keep the example code limited to just the factory definition? gets a bit long as is

Reply:
I discussed this with @dschwank and we would like to keep the example here is it is. We aligned it with the `aws-alb` provider documentation that also encompassed a full example.

Reply:
The main worry is that the maintenance burden of this is much higher than if we limit the example to the more narrow set, but then also that it takes up space and that it is a bit more difficult to figure out what changes are part of installing the auth provider vs what's just general auth plugin setup. The ALB docs are in contrib so not reviewed as thoroughly

Reply:
We reduced this example 👍